packages/protocol/contracts/stability/GrandaMento.sol

```diff
@@ -180,6 +218,134 @@ contract GrandaMento is
+    // Require that the proposal is in the Approved state.
+    require(proposal.state == ExchangeProposalState.Approved, "Proposal must be in Approved state");
+    // Require that the veto period has elapsed since the approval time.
+    require(
+      proposal.approvalTimestamp.add(vetoPeriodSeconds) <= block.timestamp,
+      "Veto period not elapsed"
+    );
+    // Mark the proposal as executed.
+    proposal.state = ExchangeProposalState.Executed;
+
+    // Perform the exchange.
+    (IERC20 sellToken, uint256 sellAmount) = getSellTokenAndSellAmount(proposal);
+    // If the exchange sells CELO, the CELO is sent to the Reserve from this contract
+    // and stable token is minted to the exchanger.
+    if (proposal.sellCelo) {
+      // Send the CELO from this contract to the reserve.
+      require(
+        sellToken.transfer(address(getReserve()), sellAmount),
+        "Transfer out of CELO to Reserve failed"
+      );
+      // Mint stable token to the exchanger.
+      require(
+        IStableToken(proposal.stableToken).mint(proposal.exchanger, proposal.buyAmount),
+        "Stable token mint failed"
+      );
+    } else {
+      // If the exchange is selling stable token, the stable token is burned from
+      // this contract and CELO is transferred from the Reserve to the exchanger.
+
+      // Burn the stable token from this contract.
+      require(IStableToken(proposal.stableToken).burn(sellAmount), "Stable token burn failed");
+      // Transfer the CELO from the Reserve to the exchanger.
+      require(
+        getReserve().transferExchangeGold(proposal.exchanger, proposal.buyAmount),
+        "Transfer out of CELO from Reserve failed"
+      );
+    }
+    emit ExchangeProposalExecuted(proposalId);
+  }
+
+  /**
+   * @notice Gets the sell token and the sell amount for a proposal.
+   * @dev For stable token sell amounts that are stored as units, the value
+   * is returned. Ensures sell amount is not greater than this contract's balance.
+   * @param proposal The proposal to get the sell token and sell amount for.
+   */
+  function getSellTokenAndSellAmount(ExchangeProposal memory proposal)
+    private
+    view
+    returns (IERC20, uint256)
+  {
+    IERC20 sellToken;
+    uint256 sellAmount;
+    if (proposal.sellCelo) {
+      sellToken = getGoldToken();
+      sellAmount = proposal.sellAmount;
+    } else {
+      address stableToken = proposal.stableToken;
+      sellToken = IERC20(stableToken);
+      // When selling stableToken, the sell amount is stored in units.
+      // Units must be converted to value when refunding.
+      sellAmount = IStableToken(stableToken).unitsToValue(proposal.sellAmount);
+    }
+    // In the event a precision issue from the unit <-> value calculations results
+    // in sellAmount being greater than this contract's balance, set the sellAmount
+    // to the entire balance.
+    // This check should not be necessary for CELO, but is done so regardless
```

Comment:
Curious to hear thoughts here

Reply:
Trying to think of an edge case but I can't think of one. To be cautious I'd try to avoid the check for CELO, but not really a strong thought.

Reply:
Thinking more about this, what would happen in the case there are many concurrent exchange proposals and we have had inflation with cUSD?

IIUC, then the ones that that are executed first will get the correct amount of stabletoken and the last one will get the remaining, effectively benefiting the ones executed first. It this an issue or just a rounding error?